### ISSUE TITLE
Negative r2

### ISSUE THREAD

**Issue by paula-tataru** [theislab/scvelo#694]
Hi,

I was investigating the results of the analysis on the pancreas example data and I noticed that the estimated r2 can be negative, with the smallest observed value when running the steady-state model to be -13.205:

```python
import scvelo as scv
import numpy as np

adata = scv.datasets.pancreas()
scv.pp.filter_and_normalize(adata, min_shared_counts=20, n_top_genes=2000)
scv.pp.moments(adata, n_neighbors=30)
scv.tl.velocity(adata, mode='deterministic')
print(np.min(adata.var["velocity_r2"]))
```

This surprised me and I investigated a bit the issue. 

Below is a paragraph from [wiki](https://en.wikipedia.org/wiki/Coefficient_of_determination) that I think might be relevant for the steady-state model:
> There are several definitions of R2 that are only sometimes equivalent. One class of such cases includes that of simple linear regression where r2 is used instead of R2. When only an intercept is included, then r2 is simply the square of the sample correlation coefficient (i.e., r) between the observed outcomes and the observed predictor values. If additional regressors are included, R2 is the square of the coefficient of multiple correlation. In both such cases, the coefficient of determination normally ranges from 0 to 1.
> 
> There are cases where the computational definition of R2 can yield negative values, depending on the definition used.

Could it be that scVelo is using a computational definition that is not suitable for the steady-state model?

I am running scVelo version 0.2.4.

Thanks,
Paula

**Comment by WeilerP** [comment]
@paula-tataru, this is not a bug. We define the R2 score as one minus the fraction of the sum of squares of residuals and total sum of squares. Consequently, R2 can become negative. A negative R2 simply tells you that your fit is worse than a horizontal line at the mean of the unspliced abundance. Do you mind sharing the phase portrait and inferred steady state ratio? Would be keen on seeing what it looks like to possible understand why the score is so low.

**Comment by paula-tataru** [comment]
The estimated gamma for the gene in question is 14.836441.
![Mbd5](https://user-images.githubusercontent.com/4553863/133208367-79ee8261-d956-4145-8d2f-f414b388b9be.png)

**Comment by WeilerP** [comment]
Thanks for sharing @paula-tataru. As you can see, the phase portrait does not have a football shape but is simply a blob. Consequently, it does not offer any insight for RNA velocity analysis (in case you expect a differentiation from green to orange/blue). You can also see that a horizontal line at approx. 1 would be a better fit compared to a straight line with the steady state ratio as slope. I guess the line is tilted to the right because of some outliers in the spliced counts.

**Comment by paula-tataru** [comment]
Hi,

I am writing on this closed issue because I think I ran into a situation where the negative r2 is throwing away valid velocity genes.

I am trying to reproduce the results of this [article](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7524549/) on a subset of the original data. The main finding in the article concerned the Lef1 gene, which was one of the top velocity genes the authors identified.

In the subset that I generated, I failed in identifying the Lef1 gene as a velocity gene and I think this can be explained by the negative r2.

```python
import scvelo as scv
import scanpy as sp

adata = sp.read_loom("GSE153596 subset.loom", validate=False)
adata.var_names_make_unique()

scv.pp.filter_and_normalize(adata, min_shared_counts=20, n_top_genes=500)
scv.pp.moments(adata, n_pcs=30, n_neighbors=30)
scv.tl.velocity(adata, mode='deterministic')

adata.var["velocity_genes"]["Lef1"]
adata.var["velocity_r2"]["Lef1"]
scv.pl.velocity(adata, ['Lef1'], save="Lef1_steady_state.png")

scv.tl.recover_dynamics(adata, n_jobs = 4)
scv.tl.velocity(adata, mode='dynamical')
adata.var["fit_likelihood"]["Lef1"]
```

Lef1 is not a velocity gene and has an r2 of about -0.38, and as a consequence is skipped in the calculations of the dynamical model and it gets assigned a nan likelihood. Arguably though, the phase portrait has the characteristic shape, which also matches quite well the figure from the article.
![Lef1_steady_state](https://user-images.githubusercontent.com/4553863/176891111-a084cb8e-41a6-4199-b6ec-d1e98fa13003.png)

Tricking the dynamical model to perform the calculations for Lef1 leads to the following
```python
adata.var["velocity_r2"]["Lef1"] = 0.9
adata.var["velocity_genes"]["Lef1"] = True
scv.tl.recover_dynamics(adata, n_jobs = 4)
scv.tl.velocity(adata, mode='dynamical')
adata.var["fit_likelihood"]["Lef1"]
scv.pl.velocity(adata, ['Lef1'], save="Lef1_dynamical.png")
```
and now Lef1 has a likelihood of about 0.26.
![Lef1_dynamical](https://user-images.githubusercontent.com/4553863/176891150-531cf30c-7a4c-46b7-b24c-a621d6e5c7b9.png)

I am running scVelo version 0.2.4.

Thanks,
Paula
